compiler/rustc_middle/src/ty/mod.rs

```diff
@@ -1252,7 +1251,11 @@ impl<'tcx> Predicate<'tcx> {
         let substs = trait_ref.skip_binder().substs;
         let pred = self.skip_binders();
         let new = pred.subst(tcx, substs);
-        if new != pred { new.potentially_quantified(tcx, PredicateKind::ForAll) } else { self }
+        if new != pred {
+            trait_ref.rebind(new).potentially_quantified(tcx, PredicateKind::ForAll)
```

Comment:
can we  use `rebind` here? I would expect that we have to union the bound vars of the `pred` and the `substs` here.

How does `T: for<'a> Foo<'a>` with `trait Foo<'a>: for<'b> Bar<'b>` work here?

Reply:
You're right. In #76814 this is actually a combination of the two. I'll change this.